### DISCUSSION TITLE
MySql——tables: Environment --- global_lsn

### DISCUSSION THREAD

**Discussion by Haohao966** [milvus-io/milvus#3774] [Q&A and General discussion] [answered]
![image](https://user-images.githubusercontent.com/70635373/93320862-24f93d80-f844-11ea-9508-0c5bd5943a7c.png)
I want to know:
What does the field name global_lsn in the Environment table in the Mysql database mean
What data will be  stored in the  Environment  table

**Comment by yhmo** [comment, marked as answer]
The global_lsn is for WAL(write ahead log) functionality.
The purpose of wal is to ensure data safety, such as insert/delete action.
Before data is serialize to disk(and record to meta service), WAL firstly store the data to local storage(under the /milvus/wal folder), if milvus crashed, the WAL can recovery the data at the next time milvus start.
So each insert/delete action has a id to WAL, we call it 'lsn'. Each collection has a latest lsn, recorded in meta system(in the Tables table). Latest lsn means the milvus has successfully consumed the insert/delete action which id equal the lsn.
For example, client invoke insert api 3 times, so there are 3 actions were recorded in WAL, their id is: 1000, 1001, 1002, 1003. Milvus successfully store the 1000/1001 action data into disk, but crashed at 1002. So the latest lsn of the collection is 1001, the global_lsn also is 1001. Then you restart milvus, the WAL will recovery the 1002 and 1003 actions.

Currently the Environment table only used to record global_lsn.

**Reply by Haohao966** [reply]
thank  you  very much

**Comment by yhmo** [comment]
If you set WAL to disable, the global_lsn will be always 0
